sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/QueryStageExec.scala

```diff
@@ -198,18 +215,23 @@ case class ShuffleQueryStageExec(
     reuse
   }
 
-  override def cancel(): Unit = shuffleFuture match {
-    case action: FutureAction[MapOutputStatistics] if !action.isCompleted =>
-      action.cancel()
-    case _ =>
+  override def cancel(): Unit = {
```

Comment:
nit: it seems better to follow `def materialize` and `def doMaterialize`, to have `def cancel` and `def doCancel`. We can put this materialization check in `def cancel`, and maybe more common logic related to cancelation in the future.

Reply:
Added `doCancel()` to `ExchangeQueryStageExec` and both QueryStageExecs implement it. Please find last commit.